release/src/router/plc-utils/lldp/respond.1
.TH respond 1 "Mar 2014" "plc-utils-2.1.6" "Qualcomm Atheros Powerline Toolkit"

.SH NAME
respond - Qualcomm Atheros Network Responder

.SH SYNOPSIS
.BR respond
.RI [ options ] 

.SH DESCRIPTION
Respond to topology requests from powerline network solicitors.
This is a proof-of-concept program that does nothing useful, as it stands.
It is maintained in case there is a need, in the future, to make it do something useful.
It requires one or more instances of program solicit to demonstrate the concept.
The underlying protocol is essentially LLDP.

.PP
This program is part of the Qualcomm Atheros Powerline Toolkit.
See the \fBplc\fR man page for an overview and installation instructions.

.SH OPTIONS

.TP
.RB - i
Select the host Ethernet interface.
All requests are sent via this host interface and only reponses received via this host interface are recognized.
The default interface is \fBeth1\fR because most people use \fBeth0\fR as their principle network connection; however, if environment string "PLC" is defined then it takes precedence over the default interface.
This option then takes precedence over either default.

.TP
-\fBp \fIprofile\fR
The configuration profile name.
The file must have two sections: "\fBStandard\fR" and "\fBSpecific\fR".
Profile elements must conform to the TLV descriptions defined in \fISimple Network Discovery Protocol Programmers Guide\fR.
The program will supply a default value for any omitted configuration profile elements.
See the example profile shown below.
The default filename is "\fBrespond.ini\fR".

.TP
.RB - q
Enter quiet mode.
Progress messages are suppressed.

.TP
.RB - v 
Enter verbose mode.
All Etherenet frames sent or received by the program are displayed on stdout.

.TP
-\fBw \fImilliseconds\fR
Wakeup frequency in milliseconds.
This is the time between announcements.
During this time, the program listens for solicitor rquests.
Values range from 0 through UINT_MAX.
'
The default is forever.
.TP
-\fB?\fR, --\fBhelp\fR
Print program help summary on stdout. This option takes precedence over other options on the command line. 
.TP
-\fB!\fR, --\fBversion\fR
Print program version information on stdout. This option takes precedence over other options on the command line. Use this option when sending screen dumps to Atheros Technical Support so that they know exactly which version of the Linux Toolkit you are using.
.SH REFERENCES
See the Qualcomm Atheros Simple Network Discovery Protocol for more information about \fBsolicitors\fR, \fBresponders\fR and \fBTLVs\fR.
.SH EXAMPLES
The following command starts a responder on the local host. The responder will announce and listen on interface \fBeth0\fR. The responder will announce values defined in profile "\fBrespond.ini\fR" by default because option \fB-p\fR is absent.
.PP
   # respond -i eth0
.PP
An example profile, "\fBrespond.ini\fR", looks something like this. This configuration file is provided for demonstration purposes only. The source code that parses this file does not support all possible variations these configuration values. See \fIIEEE Standard for Local and metworpolitan networks - Station and Meida Access Control Connectivity Discovery\fR for an explanation of the the value
.PP
   # file: respond.ini
 
   # ====================================================================
   # This is an example profile for use by program respond; it supports 
   # the Qualcomm Atheros Simple Network Discovery protocol;
   # --------------------------------------------------------------------
   
   [Standard]
   Chassis ID = 04AABBCCDDEEFF
   Port ID = 03112233445566
   Time To Live = 3600 
   Port Description = The Internet of Things
   System Name = WNC Hybrid Router
   System Description = Shanghai Public Library Wireless Network
   System Capabilities = 00:FF:FF:FF:FF
   Management Address = http://www.mamagement.com
   
   [Specific]
   Manufacturer = Qualcomm Atheros
   Hardware Model = QCA7000
   Serial Number = QCA7000-ND25B-0003
   HOST MAC = 88:77:66:55:44:33
   PLC MAC = 00:B0:52:00:BA:BE
   WIFI MAC = AA:BB:CC:DD:EE:FF
   WIFI IP = 00:FF:FF:AB:CD
   
.SH SEE ALSO
.BR plc ( 1 ),
.BR solicit ( 1 ),
.SH CREDITS
 Charles Maier <cmaier@qca.qualcomm.com>
'



release/src/router/plc-utils/lldp/respond.1.c
/*====================================================================*
*
*   Copyright (c) 2013 Qualcomm Atheros, Inc.
*
*   All rights reserved.
*
*====================================================================*/

/*====================================================================*
 *
 *   respond.c - Qualcomm Atheros Atheros Network Responder;
 *   
 *   Contributor(s):
 *      Charles Maier <cmaier@qca.qualcomm.com>
 *
 *--------------------------------------------------------------------*/

#define _GETOPT_H

/*====================================================================*
 *   system header files;
 *--------------------------------------------------------------------*/

#include <unistd.h>
#include <stdlib.h>
#include <limits.h>
#include <errno.h>

/*====================================================================*
 *   custom header files;
 *--------------------------------------------------------------------*/

#include "../tools/getoptv.h"
#include "../tools/putoptv.h"
#include "../tools/memory.h"
#include "../tools/number.h"
#include "../tools/types.h"
#include "../tools/flags.h"
#include "../tools/files.h"
#include "../tools/error.h"
#include "../tools/config.h"
#include "../ether/channel.h"
#include "../nda/lldp.h"
#include "../plc/plc.h"

/*====================================================================*
 *   custom source files;
 *--------------------------------------------------------------------*/

#ifndef MAKEFILE
#include "../tools/getoptv.c"
#include "../tools/putoptv.c"
#include "../tools/version.c"
#include "../tools/hexcopy.c"
#include "../tools/hexdump.c"
#include "../tools/hexdecode.c"
#include "../tools/hexstring.c"
#include "../tools/decdecode.c"
#include "../tools/decstring.c"
#include "../tools/uintspec.c"
#include "../tools/todigit.c"
#include "../tools/error.c"
#include "../tools/strfbits.c"
#include "../tools/config.c"
#endif

#ifndef MAKEFILE
#include "../plc/Devices.c"
#endif

#ifndef MAKEFILE
#include "../mme/EthernetHeader.c"
#include "../mme/QualcommHeader.c"
#include "../mme/UnwantedMessage.c"
#include "../mme/readmessage.c"
#include "../mme/sendmessage.c"
#endif

#ifndef MAKEFILE
#include "../ether/channel.c"
#include "../ether/openchannel.c"
#include "../ether/closechannel.c"
#include "../ether/sendpacket.c"
#include "../ether/readpacket.c"
#endif

#ifndef MAKEFILE
#include "../lldp/TLVPack.c"
#include "../lldp/TLVPackOS.c"
#include "../lldp/TLVPick.c"
#include "../lldp/TLVPeek.c"
#endif

/*====================================================================*
 *   program constants;
 *--------------------------------------------------------------------*/

#define RESPOND_VERBOSE (1 << 0)
#define RESPOND_SILENCE (1 << 1)

#define PLCDEVICE "PLC"

#define PROFILE "respond.ini"
#define SECTION "default"
#define STANDARD "Standard"
#define SPECIFIC "Specific"

/*====================================================================*
 *
 *   byte * populate1 (byte * memory, size_t extent, char const * profile, char const * section, char buffer [], size_t length) 
 *
 *   populate memory region with IEEE standard TLVs of interest to
 *   us; not all IEEE 802 TLVs are implemented here; the data source
 *   is static data declarations in one case and a configuration
 *   file in the other case;
 *
 *--------------------------------------------------------------------*/

static byte * populate1 (byte * memory, size_t extent, char const * profile, char const * section, char buffer [], size_t length) 

{ 
	char SYSTEM_NAME [HOST_NAME_MAX] = 
	{ 
		"System Name"
	}; 

#ifndef PROFILE

	byte * length = memory +  extent; 
	static uint16_t TIME_TO_LIVE = 50000; 
	static byte CHASSIS_ID [] = 
	{ 
		0x04, 
		0xAA, 
		0xBB, 
		0xCC, 
		0xDD, 
		0xEE, 
		0xFF
	}; 
	static byte PORT_ID [] = 
	{ 
		0x03, 
		0x11, 
		0x22, 
		0x33, 
		0x44, 
		0x55, 
		0x66
	}; 
	static byte SYSTEM_CAPABILITIES [] = 
	{ 
		0x00, 
		0xFF, 
		0xFF, 
		0xAB, 
		0xCD
	}; 
	static char PORT_DESCRIPTION [] = "Port Description"; 
	static char SYSTEM_DESCRIPTION [] = "System Description"; 
	gethostname (SYSTEM_NAME, sizeof (SYSTEM_NAME)); 
	memory = TLVPack (memory, length - memory, TLV_IEEE_CHASSIS_ID, sizeof (CHASSIS_ID), CHASSIS_ID); 
	memory = TLVPack (memory, length - memory, TLV_IEEE_PORT_ID, sizeof (PORT_ID), PORT_ID); 
	memory = TLVPack (memory, length - memory, TLV_IEEE_TIME_TO_LIVE, sizeof (TIME_TO_LIVE), & TIME_TO_LIVE); 
	memory = TLVPack (memory, length - memory, TLV_IEEE_PORT_DESCRIPTION, sizeof (PORT_DESCRIPTION), PORT_DESCRIPTION); 
	memory = TLVPack (memory, length - memory, TLV_IEEE_SYSTEM_NAME, strlen (SYSTEM_NAME) +  1, SYSTEM_NAME); 
	memory = TLVPack (memory, length - memory, TLV_IEEE_SYSTEM_DESCRIPTION, sizeof (SYSTEM_DESCRIPTION), SYSTEM_DESCRIPTION); 
	memory = TLVPack (memory, length - memory, TLV_IEEE_SYSTEM_CAPABILITIES, sizeof (SYSTEM_CAPABILITIES), SYSTEM_CAPABILITIES); 

#else

	uint16_t ttl; 
	char const * string; 
	byte * offset = memory; 
	memory += extent; 
	gethostname (SYSTEM_NAME, sizeof (SYSTEM_NAME)); 
	string = configstring (profile, section, "Chassis ID", "00:00:00:00:00:00:00"); 
	offset = TLVPack (offset, memory - offset, TLV_IEEE_CHASSIS_ID, hexcopy (buffer, length, string), buffer); 
	string = configstring (profile, section, "Port ID", "00:00:00:00:00:00:00"); 
	offset = TLVPack (offset, memory - offset, TLV_IEEE_PORT_ID, hexcopy (buffer, length, string), buffer); 
	string = configstring (profile, section, "Time To Live", "0"); 
	ttl = htons (atoi (string)); 
	offset = TLVPack (offset, memory - offset, TLV_IEEE_TIME_TO_LIVE, sizeof (ttl), & ttl); 
	string = configstring (profile, section, "Port Description", "Unspecified port"); 
	offset = TLVPack (offset, memory - offset, TLV_IEEE_PORT_DESCRIPTION, 1 +  strlen (string), string); 
	string = configstring (profile, section, "System Name", SYSTEM_NAME); 
	offset = TLVPack (offset, memory - offset, TLV_IEEE_SYSTEM_NAME, 1 +  strlen (string), string); 
	string = configstring (profile, section, "System Description", "Do nothing quickly and accurately"); 
	offset = TLVPack (offset, memory - offset, TLV_IEEE_SYSTEM_DESCRIPTION, 1 +  strlen (string), string); 
	string = configstring (profile, section, "System Capabilities", "00:00:00:00:00"); 
	offset = TLVPack (offset, memory - offset, TLV_IEEE_SYSTEM_CAPABILITIES, hexcopy (buffer, length, string), buffer); 
	string = configstring (profile, section, "Management URL", "http://www.management.com"); 
	offset = TLVPack (offset, memory - offset, TLV_IEEE_MANAGEMENT_ADDRESS, 1 +  strlen (string), string); 

#endif

	return (offset); 
} 

/*====================================================================*
 *
 *   byte * populate2 (byte * memory, size_t extent, char const * profile, char const * section, char buffer [], size_t length) 
 *
 *   populate memory region with IEEE organizationally specific TLVs 
 *   of interest to us; Qualcomm Atheros has defined the ones used 
 *   here; the data source here is static data declarations in one
 *   case and a configuration file in the other;
 *
 *--------------------------------------------------------------------*/

static byte * populate2 (byte * memory, size_t extent, char const * profile, char const * section, char buffer [], size_t length) 

{ 

#ifndef PROFILE

	extern struct channel channel; 
	byte * length = memory +  extent; 
	static char MANUFACTURER [] = "Qualcomm Atheros"; 
	static char HARDWARE_MODEL [] = "QCA7000"; 
	static char SERIAL_NUMBER [] = "00:B0:52:00:00:55"; 
	static byte PLC_MAC [] = 
	{ 
		0x00, 
		0xB0, 
		0x52, 
		0x00, 
		0xBA, 
		0xBE
	}; 
	static byte WIFI_MAC [] = 
	{ 
		0xAA, 
		0xBB, 
		0xCC, 
		0xDD, 
		0xEE, 
		0xFF
	}; 
	static byte WIFI_IP [] = 
	{ 
		192, 
		168, 
		1, 
		88
	}; 
	memory = TLVPackOS (memory, length - memory, TLV_OS_MANUFACTURER, sizeof (MANUFACTURER), MANUFACTURER); 
	memory = TLVPackOS (memory, length - memory, TLV_OS_HARDWARE_MODEL, sizeof (HARDWARE_MODEL), HARDWARE_MODEL); 
	memory = TLVPackOS (memory, length - memory, TLV_OS_SERIAL_NUMBER, sizeof (SERIAL_NUMBER), SERIAL_NUMBER); 
	memory = TLVPackOS (memory, length - memory, TLV_OS_HOST_MAC, sizeof (channel.host), channel.host); 
	memory = TLVPackOS (memory, length - memory, TLV_OS_WIFI_MAC, sizeof (WIFI_MAC), WIFI_MAC); 
	memory = TLVPackOS (memory, length - memory, TLV_OS_WIFI_IP, sizeof (WIFI_IP), WIFI_IP); 
	memory = TLVPackOS (memory, length - memory, TLV_OS_PLC_MAC, sizeof (PLC_MAC), PLC_MAC); 

#else

	char const * string; 
	byte * offset = memory; 
	memory += extent; 
	string = configstring (profile, section, "Manufacturer", "Unknown"); 
	offset = TLVPackOS (offset, memory - offset, TLV_OS_MANUFACTURER, 1 +  strlen (string), string); 
	string = configstring (profile, section, "Hardware Model", "Unknown"); 
	offset = TLVPackOS (offset, memory - offset, TLV_OS_HARDWARE_MODEL, 1 +  strlen (string), string); 
	string = configstring (profile, section, "Serial Number", "0000-0000-0000-0000"); 
	offset = TLVPackOS (offset, memory - offset, TLV_OS_SERIAL_NUMBER, 1 +  strlen (string), string); 
	string = configstring (profile, section, "Host MAC", "00:00:00:00:00:00"); 
	offset = TLVPackOS (offset, memory - offset, TLV_OS_HOST_MAC, hexcopy (buffer, length, string), buffer); 
	string = configstring (profile, section, "WiFi MAC", "00:00:00:00:00:00"); 
	offset = TLVPackOS (offset, memory - offset, TLV_OS_WIFI_MAC, hexcopy (buffer, length, string), buffer); 
	string = configstring (profile, section, "WiFi IP", "00:00:00:00"); 
	offset = TLVPackOS (offset, memory - offset, TLV_OS_WIFI_IP, hexcopy (buffer, length, string), buffer); 
	string = configstring (profile, section, "PLC MAC", "00:00:00:00:00:00"); 
	offset = TLVPackOS (offset, memory - offset, TLV_OS_PLC_MAC, hexcopy (buffer, length, string), buffer); 

#endif

	return (offset); 
} 

/*====================================================================*
 *
 *   byte * populate (byte * memory, size_t extent, char const * profile, char const * section);
 *
 *   populate a memory region with IEEE 802 standard TLVs then IEEE
 *   802 organizationally specific TLVs; provide the buffer that is 
 *   used to format and encode information before copying it to the
 *   memory region;
 *   
 *   customers may implement their own populate functions, possibly
 *   reading from a MIB or other type of database;
 *
 *--------------------------------------------------------------------*/

static byte * populate (byte * memory, size_t extent, char const * profile, char const * section) 

{ 
	char buffer [1024]; 
	byte * offset = memory; 
	memory += extent; 
	offset = populate1 (offset, memory - offset, profile, STANDARD, buffer, sizeof (buffer)); 
	offset = populate2 (offset, memory - offset, profile, SPECIFIC, buffer, sizeof (buffer)); 
	return (offset); 
} 

/*====================================================================*
 *
 *   void announce (struct channel * channel, struct message * message, char const * profile, char const * section, unsigned timer);
 *
 *   announce network presence and collect acknowledgements; do not 
 *   very long for an acknowledgement; timeout and start listening;
 * 
 *--------------------------------------------------------------------*/

static void announce (struct channel * channel, struct message * message, char const * profile, char const * section, unsigned timer) 

{ 

#ifndef __GNUC__
#pragma pack (push,1)
#endif

	struct __packed ms_discover_indicate 
	{ 
		struct ethernet_hdr ethernet; 
		struct qualcomm_hdr qualcomm; 
		uint8_t MDATA [1494]; 
	} 
	* indicate = (struct ms_discover_indicate *) (message); 
	struct __packed ms_discover_response 
	{ 
		struct ethernet_hdr ethernet; 
		struct qualcomm_hdr qualcomm; 
		uint8_t MDATA [1494]; 
	} 
	* response = (struct ms_discover_response *) (message); 

#ifndef __GNUC__
#pragma pack (pop)
#endif

	ssize_t length; 
	memset (message, 0, sizeof (* message)); 
	EthernetHeader (& indicate->ethernet, channel->peer, channel->host, channel->type); 
	QualcommHeader (& indicate->qualcomm, 0, (MS_DISCOVER | MMTYPE_IND)); 
	length = populate (indicate->MDATA, sizeof (indicate->MDATA), profile, section) - (byte *) (message); 
	if (length < (ETHER_MIN_LEN - ETHER_CRC_LEN)) 
	{ 
		length = (ETHER_MIN_LEN - ETHER_CRC_LEN); 
	} 
	if (sendmessage (channel, message, length) > 0) 
	{ 

#if 0

		TLVPeek (indicate->MDATA, sizeof (indicate->MDATA)); 

#endif

		while (readmessage (channel, message, 0, (MS_DISCOVER | MMTYPE_RSP)) > 0) 
		{ 
			TLVPeek (response->MDATA, sizeof (response->MDATA)); 
		} 
	} 
	return; 
} 

/*====================================================================*
 *
 *   void monitor (struct channel * channel, struct message * message, char const * profile, char const * section, unsigned timer);
 *
 *   monitor for discovery requests and respond; lengthen the channel
 *   timeout during this operations;
 * 
 *--------------------------------------------------------------------*/

static void monitor (struct channel * channel, struct message * message, char const * profile, char const * section, unsigned timer) 

{ 

#ifndef __GNUC__
#pragma pack (push,1)
#endif

	struct __packed ms_discover_request 
	{ 
		struct ethernet_hdr ethernet; 
		struct qualcomm_hdr qualcomm; 
		uint8_t MACTION; 
	} 
	* request = (struct ms_discover_request *) (message); 
	struct __packed ms_discover_confirm 
	{ 
		struct ethernet_hdr ethernet; 
		struct qualcomm_hdr qualcomm; 
		uint8_t MDATA [1494]; 
	} 
	* confirm = (struct ms_discover_confirm *) (message); 

#ifndef __GNUC__
#pragma pack (pop)
#endif

	ssize_t length; 
	unsigned timeout = channel->timeout; 
	channel->timeout = timer; 
	while (readmessage (channel, message, 0, (MS_DISCOVER | MMTYPE_REQ)) > 0) 
	{ 
		if (! request->MACTION) 
		{ 
			EthernetHeader (& confirm->ethernet, request->ethernet.OSA, channel->host, channel->type); 
			QualcommHeader (& confirm->qualcomm, request->qualcomm.MMV, (MS_DISCOVER | MMTYPE_CNF)); 
			memset (confirm->MDATA, 0, sizeof (confirm->MDATA)); 
			length = populate (confirm->MDATA, sizeof (confirm->MDATA), profile, section) - (byte *) (message); 
			if (length < (ETHER_MIN_LEN - ETHER_CRC_LEN)) 
			{ 
				length = (ETHER_MIN_LEN - ETHER_CRC_LEN); 
			} 

#if 0

			TLVPeek (confirm->MDATA, sizeof (confirm->MDATA)); 

#endif

			if (sendmessage (channel, message, length) <= 0) 
			{ 
				error (0, errno, CHANNEL_CANTSEND); 
			} 
		} 
	} 
	channel->timeout = timeout; 
	return; 
} 

/*====================================================================*
 *   
 *   int main (int argc, char * argv[]);
 *   
 *
 *--------------------------------------------------------------------*/

int main (int argc, char const * argv []) 

{ 
	extern struct channel channel; 
	extern byte const broadcast [ETHER_ADDR_LEN]; 
	static char const * optv [] = 
	{ 

#ifdef PROFILE

		"i:p:qt:vw:", 

#else

		"i:qt:vw:", 

#endif

		PUTOPTV_S_DIVINE, 
		"Qualcomm Atheros Network Responder", 

#if defined (WINPCAP) || defined (LIBPCAP) 

		"i n\thost interface is (n) [" LITERAL (CHANNEL_ETHNUMBER) "]", 

#else

		"i s\thost interface is (s) [" LITERAL (CHANNEL_ETHDEVICE) "]", 

#endif

#ifdef PROFILE

		"p s\tconfiguration profile is (s) [" LITERAL (PROFILE) "]", 

#endif

		"q\tsuppress normal output", 

#ifdef PROFILE

		"s s\tconfiguration section is (s) [" LITERAL (SECTION) "]", 

#endif

		"t n\tread timeout is (n) milliseconds [" LITERAL (CHANNEL_TIMEOUT) "]", 
		"v\tverbose messages on stdout", 
		"w n\twakeup every (n) milliseconds [" LITERAL (PLC_LONGTIME) "]", 
		(char const *) (0)
	}; 
	struct message message; 
	char const * profile = PROFILE; 
	char const * section = SECTION; 
	unsigned timer = PLC_LONGTIME; 
	signed c; 
	if (getenv (PLCDEVICE)) 
	{ 

#if defined (WINPCAP) || defined (LIBPCAP) 

		channel.ifindex = atoi (getenv (PLCDEVICE)); 

#else

		channel.ifname = strdup (getenv (PLCDEVICE)); 

#endif

	} 
	optind = 1; 
	memcpy (channel.peer, broadcast, sizeof (channel.peer)); 
	while (~ (c = getoptv (argc, argv, optv))) 
	{ 
		switch (c) 
		{ 
		case 'i': 

#if defined (WINPCAP) || defined (LIBPCAP) 

			channel.ifindex = atoi (optarg); 

#else

			channel.ifname = optarg; 

#endif

			break; 

#ifdef PROFILE

		case 'p': 
			profile = optarg; 
			break; 

#endif

		case 'q': 
			_clrbits (channel.flags, CHANNEL_SILENCE); 
			break; 

#ifdef PROFILE

		case 's': 
			section = optarg; 
			break; 

#endif

		case 't': 
			channel.timeout = (unsigned) (uintspec (optarg, 0, UINT_MAX)); 
			break; 
		case 'v': 
			_setbits (channel.flags, CHANNEL_VERBOSE); 
			break; 
		case 'w': 
			timer = (unsigned) (uintspec (optarg, 1, UINT_MAX)); 
			break; 
		default: 
			break; 
		} 
	} 
	argc -= optind; 
	argv += optind; 
	if (argc) 
	{ 
		error (1, 0, ERROR_TOOMANY); 
	} 

#ifdef PROFILE

	if (access (profile, R_OK)) 
	{ 
		error (1, errno, "Can't access '%s'", profile); 
	} 

#endif

	openchannel (& channel); 
	while (true) 
	{ 
		announce (& channel, & message, profile, section, timer); 
		monitor (& channel, & message, profile, section, timer); 
	} 
	closechannel (& channel); 
	return (0); 
} 

